mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,323 @@ def load_model(model_uri):
+
+        def clean_id(id_):
+            _get_autolog_training_status()._pred_result_id_to_dataset_name_and_run_id \
+                .pop(id_, None)
+
+        # When the `predict_result` object being GCed, its ID may be reused, so register a finalizer
+        # to clear the ID from the dict for preventing wrong ID mapping.
+        weakref.finalize(predict_result, clean_id, prediction_result_id)
+
+    @staticmethod
+    def gen_metric_call_command(self_obj, metric_fn, *call_pos_args, **call_kwargs):
+        """
+        Generate metric function call command string like `metric_fn(arg1, arg2, ...)`
+        Note: this method include inspecting argument variable name.
+         So should be called directly from the "patched method", to ensure it capture
+         correct argument variable name.
+        """
+
+        arg_list = []
+
+        def arg_to_str(arg):
+            if arg is None or np.isscalar(arg):
+                if isinstance(arg, str) and len(arg) > 32:
+                    # truncate too long string
+                    return repr(arg[:32] + '...')
+                return repr(arg)
+            else:
+                # dataset arguments or other non-scalar type argument
+                return _inspect_original_var_name(arg, fallback_name=f'<{arg.__class__.__name__}>')
+
+        param_sig = inspect.signature(metric_fn).parameters
+        param_sig_keys = list(param_sig.keys())
+
+        if self_obj is not None:
+            param_sig_keys.pop(0)
+
+        param_sig_keys = [
+            key for key in param_sig_keys
+            if param_sig[key].kind in [
+                inspect.Parameter.POSITIONAL_ONLY, inspect.Parameter.POSITIONAL_OR_KEYWORD
+            ]
+        ]
+
+        if self_obj is not None:
+            call_fn_name = f'{self_obj.__class__.__name__}.{metric_fn.__name__}'
+        else:
+            call_fn_name = metric_fn.__name__
+
+        assert len(param_sig_keys) >= len(call_pos_args), 'Found too many positional arguments.'
+        for arg_name, arg in zip(param_sig_keys, call_pos_args):
+            arg_list.append(f'{arg_name}={arg_to_str(arg)}')
+
+        for arg_name, arg in call_kwargs.items():
+            arg_list.append(f'{arg_name}={arg_to_str(arg)}')
+
+        arg_list_str = ', '.join(arg_list)
+
+        return f'{call_fn_name}({arg_list_str})'
+
+    def register_metric_info(self, run_id, metric_name, dataset_name, call_command):
+
+        call_cmd_list = self._metric_api_call_info[run_id][metric_name]
+
+        index = len(call_cmd_list)
+        metric_name_with_index = self.gen_name_with_index(metric_name, index)
+        metric_key = f'{metric_name_with_index}_{dataset_name}'
+
+        call_cmd_list.append((metric_key, call_command))
+
+        self._metric_info_artifact_need_update[run_id] = True
+        return metric_key
+
+    def register_metric_api_call(self, metric_name, call_command, call_pos_args, call_kwargs):
```

Comment:
The way these methods are structured is a bit confusing:

1. `register_metric_info` associates information about a specified metric API call with a specified `run_id`.
2. `register_metric_api_call` infers the run ID and dataset name for a metric API call and then calls `register_metric_info`.

I think that `register_metric_info` should be renamed to `register_metric_api_call` and that we should rename `register_metric_api_call` to `get_run_id_and_dataset_name_for_metric_api_call`. `get_run_id_and_dataset_name_for_metric_api_call` should just return the run ID and eval dataset name, if it exists. We should remove the `register_metric_info` call and instead have `patched_metric_api` call both of these methods:

```
                run_id, eval_dataset_name = \
                    status.get_run_id_and_dataset_name_for_metric_api_call(args, kwargs)
                if run_id and eval_dataset_name:
                    status.register_metric_info(run_id, metric_name, dataset_name, call_command)
                    status.log_eval_metric(run_id, metric_key, metric)
```